design/keycloak.x/storage-persistence.md

```diff
@@ -0,0 +1,530 @@
+     * combines the given builders with the Boolean OR operator.
+     * <p>
+     * Predicate coming out of {@code and} on an empty array of {@code builders}
+     * (i.e. empty disjunction) is always {@code false}.
+     *
+     * <pre>
+     *   cb = storage.getCriteriaBuilder();
+     *   storage.read(cb.or(
+     *     cb.compare(FIELD1, EQ, 1).compare(FIELD2, EQ, 2),
+     *     cb.compare(FIELD1, EQ, 3).compare(FIELD2, EQ, 4)
+     *   );
+     * </pre>
+     */
+    ModelCriteriaBuilder or(ModelCriteriaBuilder... builders);
+
+    /**
+     * Creates and returns a new instance of {@code ModelCriteriaBuilder} that
+     * negates the given builder.
+     * <p>
+     * Note that if the {@code builder} has no condition yet, there is nothing
+     * to negate: empty negation is always {@code true}.
+     *
+     * @param builder
+     * @return
+     */
+    ModelCriteriaBuilder not(ModelCriteriaBuilder builder);
+}
+```
+
+## <a name="_otherVersioning"></a> Other Versioning Options
+
+The versions in the [VC requirements](#_vc) are treated as sequences of integers.
+
+This is insufficient in case of long-term support releases where patches
+are issued between micro releases, but 0-downtime upgrade between minor
+releases is still a requirement.
+
+Note though that the versions in the [VC requirements](#_vc) can be
+implemented by _any_ ordered sequence. For long-term support (LTS)
+releases, this can be a pair _(minor, micro)_, where
+the [VC](#_vc) condition 1 and 3 would be defined on the whole pair,
+and the [VC](#_vc) condition 2 would be defined on the _minor_ component.
+
+To address maintainability, it is possible to impose further conditions
+for LTS version. For example, borrowing versioning from RHSSO,
+it may be valid to impose a condition that objects written
+in version 7.4.0 are readable by store from 7.3.z but objects
+from further 7.4._z_ versions (_z_ ≥ 1) may be not. Specifically,
+it means that updating from e.g. 7.3.4 to 7.4.2 is only possible
+in two separate steps: 7.3.4 &rarr; 7.4.0 and 7.4.0 &rarr; 7.4.2.
+
+## Upgrades
+
+Upgrade of a single Keycloak server in cluster should have no impact
+on the service availability. That is achieved by ensuring that
+two adjacent versions of Keycloak can be run alongside each other
+in the same cluster and share the same stores. How this is done depends
+on the storage implementation. One of the (inefficient) ways to achieve
+that is to ensure that every store version _N + 1_ writes two object
+versions - one for version _N_ and other _N + 1_. There are other ways
+possible to achieve the same, basically they count on redundancy.
+
+### Physical Schema Upgrades
+
+Some of the changes need updates of the physical storage schema. For example,
+in case of adding a single searchable field, this field should be somehow
+indexed, and in case of relational databases, that may require adding
+a column and an index. Adding an index may in turn lock the table and
```

Comment:
So does this mean that new store may still somehow support storing fields in some RDBMS like format where the indexed field will require to have it's separate columns?

If that is the case, I wonder if it would makes sense to store indexes in separate tables? For example when there is an existing table ROLE and then there will be a need to create the index on field "name", there will be a need to create new table ROLE_NAME_IDX, which will contain just the ID of the role and the NAME of it. This will mean that existing table ROLE will be still able to "serve requests" from the old version and users in the old version won't be "blocked" by adding new index?

Reply:
> So does this mean that new store may still somehow support storing fields in some RDBMS like format where the indexed field will require to have it's separate columns?

Yes, actual layout depends on the implementor.

> If that is the case, I wonder if it would makes sense to store indexes in separate tables? For example when there is an existing table ROLE and then there will be a need to create the index on field "name", there will be a need to create new table ROLE_NAME_IDX, which will contain just the ID of the role and the NAME of it. This will mean that existing table ROLE will be still able to "serve requests" from the old version and users in the old version won't be "blocked" by adding new index?

Yes, this may be an option. Yet another option may be to leverage database capable of building indices on JSON expressions, e.g. https://bitnine.net/blog-postgresql/postgresql-internals-jsonb-type-and-its-indexes/

Reply:
Cool, I wasn't aware of this option. I am sure that you will eventually figure the most optimal way for each RDBMS :-) Just wanted to clearify this a bit...